### PR TITLE
gpu: Add OpRayQueryInitializeKHR validation

### PR DESCRIPTION
**Pull request by spencer-lunarg**
This adds some (will add more after we confirm this flow works) SPIR-V instrumentation to validate the values of `OpRayQueryInitializeKHR` instructions

Positive versions of tests are in https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/7520

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -143,8 +144,6 @@ target_sources(vk_layer_validation_tests PRIVATE
     unit/query_positive.cpp
     unit/ray_tracing.cpp
     unit/ray_tracing_nv.cpp
-    unit/ray_tracing_gpu_nv.cpp
```
Comment (spencer-lunarg) [inline tests/CMakeLists.txt:146]: this was removed in https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/7595 already, will be rebased

File: layers/gpu_validation/spirv/ray_query_pass.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+// Create a pass to instrument SPV_KHR_ray_query instructions
+class RayQueryPass : public Pass {
+  public:
+    RayQueryPass(Module& module) : Pass(module) {}
+
+  private:
+    bool AnalyzeInstruction(const Function& function, const Instruction& inst) override;
+    uint32_t CreateFunctionCall(BasicBlock& block) override;
+    void Reset() override;
```
Comment (arno-lunarg) [inline layers/gpu_validation/spirv/ray_query_pass.h:35]: Nit: could use `final` instead of `override`

File: tests/unit/gpu_av_ray_query.cpp [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        layout(set = 0, binding = 1) uniform Uniforms {
+          float t_min;
+          float t_max;
+        } trace_rays_params;
+
+        void main() {
+            rayQueryEXT query;
+            rayQueryInitializeEXT(query, tlas, gl_RayFlagsTerminateOnFirstHitEXT, 0xff, vec3(0), 
+              trace_rays_params.t_min, vec3(0,0,1), trace_rays_params.t_max);
+            rayQueryProceedEXT(query);
```
Comment (arno-lunarg) [inline tests/unit/gpu_av_ray_query.cpp:43]: So as discussed we need to make sure that passing down a null query to the driver is ok

File: layers/gpu_validation/gpu_error_message.cpp [hunk 4]
```diff
@@ -436,6 +436,60 @@ bool gpuav::Validator::GenerateValidationMessage(const uint32_t *debug_record, c
                     assert(false);
             }
         } break;
+        case kInstErrorRayQueryNegativeMin: {
```
Comment (arno-lunarg) [inline layers/gpu_validation/gpu_error_message.cpp:439]: I wanted to get rid of this `GenerationValidationMessage`, now even more so. Same for `gpu_shader_constants.h`, keeping adding new constants like so is not sustainable.
But it is ok for this PR!

**Review by arno-lunarg — APPROVED** [review]
I will not buy any more french pastries unless you add the missing new lines at the end of files

More seriously, have you tried running this against an application using ray query? Maybe the ray query sample for instance

File: layers/gpu_validation/gpu_error_message.cpp (hunk 4 above)
Comment (spencer-lunarg) [reply layers/gpu_validation/gpu_error_message.cpp:439]: ya, 100% agree, was going to be something I asked you tomorrow lol

File: tests/unit/gpu_av_ray_query.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply tests/unit/gpu_av_ray_query.cpp:43]: I still need to double check the spec (also plan to run against the samples)

**Comment by jeremyg-lunarg** [comment]
> I will not buy any more french pastries unless you add the missing new lines at the end of files

BRB, writing a PR to add all the missing EOF new lines.

File: layers/gpu_shaders/inst_ray_query.comp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        if (opaque_mask != 0 && ((opaque_mask & (opaque_mask - 1)) != 0)) {
+            error = kInstErrorRayQueryOpaque;
+            param_0 = ray_flags;
+            break;
+        }
+    } while (false);
+
+    if (0u != error) {
+        uint rec_len = 9u;
```
Comment (jeremyg-lunarg) [inline layers/gpu_shaders/inst_ray_query.comp:103]: does the GLSL preprocessor support function-like macros? I know that making this a function has been problematic but it is slightly disturbing to have it inline everywhere.

File: layers/gpu_validation/gpu_error_message.cpp [hunk 6]
```diff
@@ -436,6 +436,60 @@ bool gpuav::Validator::GenerateValidationMessage(const uint32_t *debug_record, c
...
+            out_vuid_msg = "VUID-RuntimeSpirv-OpRayQueryInitializeKHR-06890";
+            error_found = true;
+        } break;
+        case kInstErrorRayQueryOpaque: {
+            const uint32_t value = debug_record[kInstRayQueryOutParam0];
+            strm << "OpRayQueryInitializeKHR operand Ray Flags is 0x" << std::hex << value << ". ";
+            out_vuid_msg = "VUID-RuntimeSpirv-OpRayQueryInitializeKHR-06891";
+            error_found = true;
+        } break;
 #if 0
```
Comment (jeremyg-lunarg) [inline layers/gpu_validation/gpu_error_message.cpp:493]: #if 0???

**Review by jeremyg-lunarg — APPROVED** [review]

File: layers/gpu_shaders/inst_ray_query.comp (hunk 5 above)
Comment (spencer-lunarg) [reply layers/gpu_shaders/inst_ray_query.comp:103]: we have GLSL experts around, we should ask

File: layers/gpu_validation/gpu_error_message.cpp (hunk 6 above)
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_error_message.cpp:493]: Oh this is me, a leftover, will remove it in a subsequent PR

Comment (spencer-lunarg) [reply layers/gpu_validation/gpu_error_message.cpp:493]: I need to rebase this, so can just make an additional commit and remove it

File: tests/unit/gpu_av_ray_query.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply tests/unit/gpu_av_ray_query.cpp:43]: Ran the samples, this is working fine, last thing I want to do is still double check the spec what can be done if we find this instruction is bad

**Pull request merged**